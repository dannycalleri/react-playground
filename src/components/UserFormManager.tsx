import React, { useEffect, useRef, useState } from "react";
import classNames from "classnames";
import { UserForm } from "./UserForm";
import { useOnClickOutside } from "../hooks/useOnClickOutside";
import type { User, StateProps } from "../types";

import styles from "./UserFormManager.module.scss";

interface UserFormManagerProps extends StateProps {
  title: string;
  name?: string;
  friends?: number[];
  cleanUpFieldsOnSave?: boolean;
  mainWindowFunction: (name: string, friends: number[]) => Promise<User>;
  newUserFunction: (name: string, friends: number[]) => Promise<User>;
}

export function UserFormManager(props: UserFormManagerProps) {
  const {
    title,
    friends = [],
    name = "",
    cleanUpFieldsOnSave = true,
    mainWindowFunction,
    newUserFunction,
  } = props;
  const topmostWindowRef = useRef<HTMLDivElement>(null);
  const [saveOrAbort, setSaveOrAbort] = useState(false);
  const [windows, setWindows] = useState<{ friends: number[] }[]>([
    { friends },
  ]);

  function handleSave(id: number) {
    console.log(`id of the user just saved: ${id}`);

    setSaveOrAbort(false);

    if (windows.length > 1) {
      const newWindows = [...windows.slice(0, windows.length - 1)];
      const lastWindow = newWindows[newWindows.length - 1];
      if (lastWindow) {
        lastWindow.friends = [...lastWindow.friends, id];
      }
      setWindows(newWindows);
    }
  }

  function openNewFriendWindow() {
    setWindows([...windows, { friends: [] }]);
  }

  function handleAbort() {
    setSaveOrAbort(false);
    setWindows([...windows.slice(0, windows.length - 1)]);
  }

  useOnClickOutside(topmostWindowRef, () => {
    setSaveOrAbort(true);
  });

  useEffect(() => {
    if (friends.length > 0) {
      setWindows((windows) => {
        const newWindow = { ...windows[0] };
        newWindow.friends = friends;
        return [newWindow, ...windows.slice(1)];
      });
    }
  }, [friends]);

  return (
    <div className={styles["windows-container"]}>
      {windows.map((window, i) => {
        const isWindowTopmost =
          windows.length === 1 || i === windows.length - 1;
        const isFirstWindow = i === 0;

        return (
          <div
            ref={
              isWindowTopmost && windows.length > 1 ? topmostWindowRef : null
            }
            className={classNames(styles.window, {
              [styles.disabled]: !isWindowTopmost,
            })}
            key={i}
            role="dialog"
          >
            <UserForm
              title={isFirstWindow ? title : "New User"}
              cleanUpFieldsOnSave={cleanUpFieldsOnSave}
              saveFunction={
                isFirstWindow ? mainWindowFunction : newUserFunction
              }
              name={isFirstWindow ? name : ""}
              friends={window.friends}
              disabled={!isWindowTopmost}
              onSave={handleSave}
              onAbort={handleAbort}
              openNewFriendWindow={openNewFriendWindow}
              saveOrAbort={isWindowTopmost && windows.length > 1 && saveOrAbort}
              state={props.state}
            />
          </div>
        );
      })}
    </div>
  );
}
