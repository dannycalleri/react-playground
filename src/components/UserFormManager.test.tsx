import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { UserFormManager } from "./UserFormManager";

describe("UserFormManager", () => {
  it("should render correctly", () => {
    const { asFragment } = render(
      <UserFormManager
        title="This is a title"
        mainWindowFunction={jest.fn()}
        newUserFunction={jest.fn()}
        state={{ data: [{ id: 1, name: "Sandro" }], dispatch: jest.fn() }}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should open a new window correctly", () => {
    const mainWindowFunction = jest.fn();
    const newUserFunction = jest.fn();
    const dispatch = jest.fn();
    render(
      <UserFormManager
        title="This is a title"
        mainWindowFunction={mainWindowFunction}
        newUserFunction={newUserFunction}
        state={{ data: [{ id: 1, name: "Sandro" }], dispatch: dispatch }}
      />
    );

    userEvent.click(
      screen.getByRole("button", {
        name: /new friend/i,
      })
    );

    expect(screen.getAllByRole("dialog")).toHaveLength(2);
  });

  it("should ask user to save or abort when clicking outside the dialog", () => {
    const mainWindowFunction = jest.fn();
    const newUserFunction = jest.fn();
    const dispatch = jest.fn();
    render(
      <UserFormManager
        title="This is a title"
        mainWindowFunction={mainWindowFunction}
        newUserFunction={newUserFunction}
        state={{ data: [{ id: 1, name: "Sandro" }], dispatch: dispatch }}
      />
    );

    userEvent.click(
      screen.getByRole("button", {
        name: /new friend/i,
      })
    );
    userEvent.click(screen.getAllByRole("dialog")[0]);

    expect(
      screen.getByText(/save or abort the current operation\?/i)
    ).toBeDefined();
  });
});
