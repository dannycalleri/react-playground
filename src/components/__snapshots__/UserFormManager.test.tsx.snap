// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserFormManager should render correctly 1`] = `
<DocumentFragment>
  <div
    class="windows-container"
  >
    <div
      class="window"
      role="dialog"
    >
      <h1>
        This is a title
      </h1>
      <form>
        <fieldset>
          <label
            for="name"
          >
            Name
          </label>
          <input
            id="name"
            type="text"
            value=""
          />
          <div>
            <div>
              <button
                type="button"
              >
                Select friend
              </button>
              <button
                type="button"
              >
                New friend
              </button>
            </div>
          </div>
          <div>
            <button
              type="submit"
            >
              Save
            </button>
          </div>
           
        </fieldset>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
